dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp

```diff
@@ -0,0 +1,462 @@
+        IN OUT LPWSTR pszFullName);
+    virtual ~RecycleBin5File();
+
+    /* IUnknown methods */
+    STDMETHODIMP QueryInterface(REFIID riid, void **ppvObject) override;
+    STDMETHODIMP_(ULONG) AddRef() override;
+    STDMETHODIMP_(ULONG) Release() override;
+
+    /* IRecycleBinFile methods */
+    STDMETHODIMP GetLastModificationTime(FILETIME *pLastModificationTime) override;
+    STDMETHODIMP GetDeletionTime(FILETIME *pDeletionTime) override;
+    STDMETHODIMP GetFileSize(ULARGE_INTEGER *pFileSize) override;
+    STDMETHODIMP GetPhysicalFileSize(ULARGE_INTEGER *pPhysicalFileSize) override;
+    STDMETHODIMP GetAttributes(DWORD *pAttributes) override;
+    STDMETHODIMP GetFileName(SIZE_T BufferSize, LPWSTR Buffer, SIZE_T *RequiredSize) override;
+    STDMETHODIMP GetTypeName(SIZE_T BufferSize, LPWSTR Buffer, SIZE_T *RequiredSize) override;
+    STDMETHODIMP Delete() override;
+    STDMETHODIMP Restore() override;
+
+protected:
+    LONG m_ref;
+    IRecycleBin5 *m_recycleBin;
+    DELETED_FILE_RECORD m_deletedFile;
+    LPWSTR m_FullName;
+};
+
+STDMETHODIMP RecycleBin5File::QueryInterface(REFIID riid, void **ppvObject)
+{
+    TRACE("(%p, %s, %p)\n", this, debugstr_guid(&riid), ppvObject);
+
+    if (!ppvObject)
+        return E_POINTER;
+
+    if (IsEqualIID(riid, IID_IUnknown) || IsEqualIID(riid, IID_IRecycleBinFile))
+        *ppvObject = static_cast<IRecycleBinFile *>(this);
+    else if (IsEqualIID(riid, IID_IExtractIconA) || IsEqualIID(riid, IID_IExtractIconW))
+    {
+        DWORD dwAttributes;
+        if (GetAttributes(&dwAttributes) == S_OK)
+            return SHCreateFileExtractIconW(m_FullName, dwAttributes, riid, ppvObject);
+        else
+            return S_FALSE;
+    }
+    else
+    {
+        *ppvObject = NULL;
+        return E_NOINTERFACE;
+    }
+
+    AddRef();
+    return S_OK;
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5File::AddRef()
+{
+    TRACE("(%p)\n", this);
+    return InterlockedIncrement(&m_ref);
+}
+
+RecycleBin5File::~RecycleBin5File()
+{
+    TRACE("(%p)\n", this);
+    m_recycleBin->Release();
+    SHFree(m_FullName);
+}
+
+STDMETHODIMP_(ULONG) RecycleBin5File::Release()
+{
+    TRACE("(%p)\n", this);
+    ULONG refCount = InterlockedDecrement(&m_ref);
+    if (refCount == 0)
+        delete this;
+    return refCount;
+}
+
+STDMETHODIMP RecycleBin5File::GetLastModificationTime(FILETIME *pLastModificationTime)
+{
+    TRACE("(%p, %p)\n", this, pLastModificationTime);
+
+    DWORD dwAttributes = ::GetFileAttributesW(m_FullName);
+    if (dwAttributes == INVALID_FILE_ATTRIBUTES)
+        return HRESULT_FROM_WIN32(GetLastError());
+
+    HANDLE hFile;
+    if (dwAttributes & FILE_ATTRIBUTE_DIRECTORY)
+        hFile = CreateFileW(m_FullName, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    else
+        hFile = CreateFileW(m_FullName, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
```

Comment:
Something along these lines, maybe?
```suggestion
    hFile = CreateFileW(m_FullName,
                        GENERIC_READ,
                        FILE_SHARE_READ |
                            ((dwAttributes & FILE_ATTRIBUTE_DIRECTORY) ? (FILE_SHARE_WRITE | FILE_SHARE_DELETE) : 0),
                        NULL,
                        OPEN_EXISTING,
                        (dwAttributes & FILE_ATTRIBUTE_DIRECTORY) ? FILE_FLAG_BACKUP_SEMANTICS : 0,
                        NULL);
```